shared-bindings/bitmapfilter/__init__.c

```diff
@@ -0,0 +1,578 @@
+//|     """A weight object to use with mix()"""
+//|
+//|     def __init__(
+//|         self,
+//|         rr: float,
+//|         rg: float,
+//|         rb: float,
+//|         gr: float,
+//|         gg: float,
+//|         gb: float,
+//|         br: float,
+//|         bg: float,
+//|         bb: float,
+//|     ) -> None:
+//|         """Perform a mixing operation where each channel can receive a fraction of every other channel.
+//|
+//|         The parameters with names like ``rb`` give the fraction of
+//|         each channel to mix into every other channel. For instance,
+//|         ``rb`` gives the fraction of blue to mix into red, and ``gg``
+//|         gives the fraction of green to mix into green."""
+//|
+static const mp_obj_namedtuple_type_t bitmapfilter_channel_mixer_type = {
+    NAMEDTUPLE_TYPE_BASE_AND_SLOTS(MP_QSTR_ChannelMixer),
+    .n_fields = 9,
+    .fields = {
+        MP_QSTR_rr,
+        MP_QSTR_rg,
+        MP_QSTR_rb,
+        MP_QSTR_gr,
+        MP_QSTR_gg,
+        MP_QSTR_gb,
+        MP_QSTR_br,
+        MP_QSTR_bg,
+        MP_QSTR_bb,
+    },
+};
+
+//| class ChannelMixerOffset:
+//|     """A weight object to use with mix()"""
+//|
+//|     def __init__(
+//|         self,
+//|         rr: float,
+//|         rg: float,
+//|         rb: float,
+//|         r_add: float,
+//|         gr: float,
+//|         gg: float,
+//|         gb: float,
+//|         g_add: float,
+//|         br: float,
+//|         bg: float,
+//|         bb: float,
+//|         b_add: float,
+//|     ) -> None:
+//|         """Perform a mixing operation where each channel can receive a fraction of every other channel, plus an offset value.
+//|
+//|         The parameters with names like ``rb`` give the fraction of
+//|         each channel to mix into every other channel. For instance,
+//|         ``rb`` gives the fraction of blue to mix into red, and ``gg``
+//|         gives the fraction of green to mix into green. The ``_add``
+//|         parameters give an offset value to add to the channel."""
+//|
+static const mp_obj_namedtuple_type_t bitmapfilter_channel_mixer_offset_type = {
+    NAMEDTUPLE_TYPE_BASE_AND_SLOTS(MP_QSTR_ChannelMixerOffset),
+    .n_fields = 9,
+    .fields = {
+        MP_QSTR_rr,
+        MP_QSTR_rg,
+        MP_QSTR_rb,
+        MP_QSTR_r_add,
+        MP_QSTR_gr,
+        MP_QSTR_gg,
+        MP_QSTR_gb,
+        MP_QSTR_g_add,
+        MP_QSTR_br,
+        MP_QSTR_bg,
+        MP_QSTR_bb,
+        MP_QSTR_b_add,
+    },
+};
+
+//| def mix(
+//|     bitmap: displayio.Bitmap, weights: Sequence[int], mask: displayio.Bitmap | None = None
+//| ) -> displayio.Bitmap:
+//|     """Perform a channel mixing operation on the bitmap
+//|
+//|     This is similar to the "channel mixer" tool in popular photo editing software.
+//|
+//|     The ``bitmap``, which must be in RGB565_SWAPPED format, is modified
+//|     according to the ``weights``.
+//|
+//|     If ``weights`` is a list of length 3 (or a `ChannelScale`
```

Comment:
I'd enforce using the named tuples because we're using the code space to make them possible. Otherwise code can use the less readable sequence version.

Reply:
I didn't enforce this but I changed the docs to list the specific types instead of Sequence[int].

Reply:
Is there example code that uses plain sequences?

Reply:
there is code in the tests that should be revised.

Reply:
I went ahead and made it require the correct types, and fixed the tests.